### PR TITLE
driver/sensor: lsm6dso: Add support

### PR DESCRIPTION
**Pull request by avisconti**
Add support to STM LSM6DSO 6-axis IMU sensor driver.
This driver supports communication with device though both
I2C and SPI bus and both polling and drdy trigger mode.

This driver supports also the sensorhub mode with the possibility
to connect a maximum of two external devices, typically a
magnetometer and an environmental sensor, currently selected among
following devices: lis2mdl magnetometer, lps22hh or lps22hb
pressure and temperature sensors, HTS221 humidity sensor.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

Note:
It makes use of stmemsc HAL i/f.
It replaces #13478

### CODE REVIEW

**Comment by avisconti** [comment]
@MaureenHelm 
I applied here all the things highlighted by you for lps22hh.

File: dts/bindings/sensor/st,lsm6dso-i2c.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    !include i2c-device.yaml
+
+properties:
+    compatible:
+        constraint: "st,lsm6dso"
+
+    irq-gpios:
+      type: compound
+      category: required
+      generation: define, use-prop-name
```
Comment (MaureenHelm) [inline dts/bindings/sensor/st,lsm6dso-i2c.yaml]: Remove `use-prop-name`

File: dts/bindings/sensor/st,lsm6dso-spi.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    !include spi-device.yaml
+
+properties:
+    compatible:
+        constraint: "st,lsm6dso"
+
+    irq-gpios:
+      type: compound
+      category: required
+      generation: define, use-prop-name
```
Comment (MaureenHelm) [inline dts/bindings/sensor/st,lsm6dso-spi.yaml]: Remove `use-prop-name`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/st,lsm6dso-i2c.yaml (hunk 1 above)
Comment (avisconti) [reply dts/bindings/sensor/st,lsm6dso-i2c.yaml]: oops... forgot it.

File: dts/bindings/sensor/st,lsm6dso-spi.yaml (hunk 2 above)
Comment (avisconti) [reply dts/bindings/sensor/st,lsm6dso-spi.yaml]: ditto

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+		LOG_DBG("failed to set BDU mode");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static struct lsm6dso_data lsm6dso_data;
+
+static struct lsm6dso_config lsm6dso_config = {
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/lsm6dso.c]: ok

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+	.sample_fetch = lsm6dso_sample_fetch,
+	.channel_get = lsm6dso_channel_get,
+};
+
+static int lsm6dso_init_chip(struct device *dev)
+{
+	struct lsm6dso_data *lsm6dso = dev->driver_data;
+	u8_t chip_id;
+
+	#if 0
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/lsm6dso.c]: ok

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+				     struct lsm6dso_data *data)
+{
+	return lsm6dso_accel_get_channel(chan, val, data,
+					data->acc_gain);
+}
+
+static inline void lsm6dso_gyro_convert(struct sensor_value *val, int raw_val,
+					float sensitivity)
+{
+	double dval;
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/lsm6dso.c]: fixed

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 6]
```diff
@@ -0,0 +1,842 @@
...
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static inline void lsm6dso_accel_convert(struct sensor_value *val, int raw_val,
+					 float sensitivity)
+{
+	double dval;
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/lsm6dso.c]: fixed

File: drivers/sensor/lsm6dso/Kconfig [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	bool "Enable LPS22HH as external sensor"
+	default y
+
+config LSM6DSO_EXT_HTS221
+	bool "Enable HTS221 as external sensor"
+	default n
+
+config LSM6DSO_EXT_LPS22HB
+	bool "Enable LPS22HB as external sensor"
+	default n
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/Kconfig]: ok

File: drivers/sensor/lsm6dso/Kconfig [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	bool "Enable LIS2MDL as external sensor"
+	default y
+
+config LSM6DSO_EXT_LPS22HH
+	bool "Enable LPS22HH as external sensor"
+	default y
+
+config LSM6DSO_EXT_HTS221
+	bool "Enable HTS221 as external sensor"
+	default n
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/Kconfig]: ok

File: drivers/sensor/lsm6dso/Kconfig [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+endif # LSM6DSO_TRIGGER
+
+config LSM6DSO_ENABLE_TEMP
+	bool "Enable temperature"
+	help
+	  Enable/disable temperature
+
+config LSM6DSO_SENSORHUB
+	bool "Enable I2C sensorhub feature"
+	default n
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/Kconfig]: ok

File: drivers/sensor/lsm6dso/Kconfig [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+	  Priority of thread used by the driver to handle interrupts.
+
+config LSM6DSO_THREAD_STACK_SIZE
+	int "Thread stack size"
+	depends on LSM6DSO_TRIGGER_OWN_THREAD
+	default 1024
+	help
+	  Stack size of thread used by the driver to handle interrupts.
+
+choice
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lsm6dso/Kconfig]: I agree, thx

**Comment by avisconti** [comment]
@MaureenHelm 
Applied all the changes, re-tested in all modes, re-pushed.
thx

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**